DaemonSet: resync daemonsets when node status has changed

cezarsa opened the pull request:
This PR changes the DaemonSet controller to also consider changes to node status to resync daemons. Currently only changes to node taints and labels are considered otherwise the daemons won't be resynced. 

It's important to check for changes in status because transient issues like insufficient allocatable resources and conditions like NodeOutOfDisk would forever prevent a daemon to be spawned on a node. Once that happened the only way to have daemons on previously neglected nodes would be recreating the daemonset or restarting the controller manager.

```release-note
```

cezarsa committed:
DaemonSet: resync daemonsets when node status has changed

This is important because a daemon set may not be scheduled on a node at
first because of things like insufficient resources. However, once there
are sufficient resources again it's expected that the daemon will be
spawned on nodes.

mikedanese commented:
Status is update every 10 seconds as part of kubelet hearbeat.

mikedanese commented:
We are in the process of moving node conditions to taints. I think that solves this problem. cc @davidopp

cezarsa commented:
Would checking for changes to `node.Status.Conditions` and
`node.Status.Allocatable` be an acceptable change to avoid the resync every
10 seconds?

Even moving conditions to taints wouldn't help with changes to allocatable
resources. This is specially troubling when unreachable node resources are
created as they would start with zeroed allocatable resources and would
never receive daemons even after becoming reachable.

On May 23, 2017 10:18 PM, "Mike Danese" <notifications@github.com> wrote:

> Status is update every 10 seconds as part of kubelet hearbeat.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/46307#issuecomment-303579133>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAArIWqQlIe5_n0shG76qgG3bhIq7BCnks5r83hwgaJpZM4NkRS6>
> .
>

cezarsa commented:
Oh sorry, I just realized the timestamps are inside the conditions so my last suggestion doesn't really help. I'll take a look at ignoring them and see what I can come up tomorrow.

cezarsa commented:
Just updated this with the suggestions. Now only differences to`node.Status.Allocatable` and the `.Type` and `.Status` fields of `NodeCondition`'s are considered. This should avoid resyncing too frequently. I also added a couple more of test situations to validate this change.

mikedanese commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -425,11 +425,28 @@ func (dsc *DaemonSetsController) addNode(obj interface{}) {
+	if len(oldConditions) != len(curConditions) {
+		return false
+	}
+	for i := range oldConditions {
```
Is the order of this guaranteed?

cezarsa replied:
I guessed it was, otherwise the reflect.DeepEqual approach would also be flawed. However, I'm not really sure, from taking a look at nodes in my cluster it seems like it. I could add a sort call before iterating just to be sure.

kargakis commented:
Duplicate of https://github.com/kubernetes/kubernetes/pull/45649

cezarsa commented:
Oops I really missed #45649 when searching for something similar. However, it only checks for changes to `Conditions` but ignores changes to `Allocatable`. Would it be okay for me to wait for it to be merged and them update this PR to only add the `Allocatable` check?

kargakis commented:
@cezarsa fine by me

kargakis commented:
/release-note-none

kargakis commented:
@k8s-bot ok to test

k82cn commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -425,11 +425,28 @@ func (dsc *DaemonSetsController) addNode(obj interface{}) {
-		// If node labels and taints didn't change, we can ignore this update.
+	if reflect.DeepEqual(oldNode.Labels, curNode.Labels) &&
+		reflect.DeepEqual(oldNode.Spec.Taints, curNode.Spec.Taints) &&
+		reflect.DeepEqual(oldNode.Status.Allocatable, curNode.Status.Allocatable) &&
```
Can you share which we expect allocatable changed? The pod start/stop on the node will not change it.

cezarsa replied:
I was thinking more about situations where the kubelet process would be restarted with a different `--max-pods` or `--system-reserved` flags. Also, some virtual machine providers support 'hot plugging' more memory/cpu to a live machine, requiring only a kubelet restart to have these values updated.
However, if it's guaranteed that the node's `Ready` condition will flip between True and False on every kubelet restart, then there's no need for this. But I wasn't able to verify that this is always the case.
